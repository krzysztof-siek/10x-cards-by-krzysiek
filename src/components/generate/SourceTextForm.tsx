import { useState } from 'react';
import type { FormEvent, ChangeEvent } from 'react';
import { Button } from '../ui/button';
import { Textarea } from '../ui/textarea';
import type { GenerateFlashcardsCommand } from '../../types';

interface SourceTextFormProps {
  isLoading: boolean;
  onSubmit: (data: GenerateFlashcardsCommand) => void;
}

export const SourceTextForm = ({ 
  isLoading, 
  onSubmit 
}: SourceTextFormProps) => {
  const [sourceText, setSourceText] = useState('');
  
  // Validation constraints
  const MIN_LENGTH = 1000;
  const MAX_LENGTH = 10000;
  const textLength = sourceText.length;
  const isTextValid = textLength >= MIN_LENGTH && textLength <= MAX_LENGTH;
  
  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (isTextValid && !isLoading) {
      onSubmit({ source_text: sourceText });
    }
  };

  return (
    <form onSubmit={handleSubmit} className="space-y-4">
      <div>
        <Textarea
          placeholder="Wklej swój tekst tutaj (1 000-10 000 znaków)..."
          value={sourceText}
          onChange={(e: ChangeEvent<HTMLTextAreaElement>) => setSourceText(e.target.value)}
          rows={10}
          className="w-full resize-none font-mono text-sm"
          disabled={isLoading}
        />
        
        <div className="flex justify-between text-sm mt-1">
          <div className={textLength > 0 && !isTextValid ? "text-red-500" : "text-gray-500"}>
            {textLength} / {MAX_LENGTH} znaków
          </div>
          {textLength > 0 && textLength < MIN_LENGTH && (
            <div className="text-amber-600">
              Wymagane minimum {MIN_LENGTH} znaków
            </div>
          )}
          {textLength > MAX_LENGTH && (
            <div className="text-red-600">
              Tekst jest zbyt długi
            </div>
          )}
        </div>
      </div>
      
      <Button 
        type="submit"
        disabled={!isTextValid || isLoading}
        className="w-full"
      >
        {isLoading ? 'Generowanie...' : 'Generuj fiszki'}
      </Button>
    </form>
  );
}; 