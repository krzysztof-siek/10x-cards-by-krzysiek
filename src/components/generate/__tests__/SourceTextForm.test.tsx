import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi, describe, it, expect, beforeEach } from "vitest";
import { SourceTextForm } from "../SourceTextForm";

describe("SourceTextForm", () => {
  const mockSubmit = vi.fn();

  beforeEach(() => {
    vi.resetAllMocks();
  });

  it("renders the form with textarea and button", () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    expect(screen.getByPlaceholderText(/wklej swój tekst tutaj/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /generuj fiszki/i })).toBeInTheDocument();
  });

  it("disables the button when text is too short", async () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    const shortText = "This is a short text.";

    await userEvent.type(textarea, shortText);

    const submitButton = screen.getByRole("button", { name: /generuj fiszki/i });
    expect(submitButton).toBeDisabled();
  });

  it("disables the button when text is too long", async () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    // Use fireEvent instead of userEvent for large text to avoid timeouts
    fireEvent.change(textarea, { target: { value: "a".repeat(11000) } });

    const submitButton = screen.getByRole("button", { name: /generuj fiszki/i });
    expect(submitButton).toBeDisabled();
  });

  it("enables the button when text is within the valid range", async () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    // Use fireEvent instead of userEvent for large text to avoid timeouts
    fireEvent.change(textarea, { target: { value: "a".repeat(5000) } });

    const submitButton = screen.getByRole("button", { name: /generuj fiszki/i });
    expect(submitButton).not.toBeDisabled();
  });

  it("displays character count correctly", async () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    fireEvent.change(textarea, { target: { value: "Hello world" } });

    // Check if the character count text is updated
    expect(screen.getByText("11 / 10000 znaków")).toBeInTheDocument();
  });

  it("calls onSubmit with correct data when form is submitted", async () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    const validText = "a".repeat(1000);

    fireEvent.change(textarea, { target: { value: validText } });

    // Find the form element using container
    const form = document.querySelector("form");
    fireEvent.submit(form as HTMLElement);

    expect(mockSubmit).toHaveBeenCalledWith({ source_text: validText });
  });

  it("disables inputs and shows loading state when isLoading is true", () => {
    render(<SourceTextForm isLoading={true} onSubmit={mockSubmit} />);

    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    expect(textarea).toBeDisabled();

    const submitButton = screen.getByRole("button", { name: /generowanie/i });
    expect(submitButton).toBeDisabled();
    expect(submitButton).toHaveTextContent("Generowanie...");
  });

  it("prevents form submission when the form is invalid", async () => {
    render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);

    // Form with too short text should not submit
    const textarea = screen.getByPlaceholderText(/wklej swój tekst tutaj/i);
    await userEvent.type(textarea, "short text");

    // Find the form element using container
    const form = document.querySelector("form");
    fireEvent.submit(form as HTMLElement);

    expect(mockSubmit).not.toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    const { container } = render(<SourceTextForm isLoading={false} onSubmit={mockSubmit} />);
    expect(container).toMatchSnapshot();
  });
});
