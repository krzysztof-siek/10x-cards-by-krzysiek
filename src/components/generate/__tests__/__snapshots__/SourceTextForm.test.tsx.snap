// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SourceTextForm > should match snapshot 1`] = `
<div>
  <form
    class="space-y-4"
  >
    <div>
      <textarea
        class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 rounded-md border bg-transparent px-3 py-2 shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm w-full resize-none font-mono text-sm"
        data-slot="textarea"
        placeholder="Wklej swój tekst tutaj (1 000-10 000 znaków)..."
        rows="10"
      />
      <div
        class="flex justify-between text-sm mt-1"
      >
        <div
          class="text-gray-500"
        >
          0
           / 
          10000
           znaków
        </div>
      </div>
    </div>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
      data-slot="button"
      disabled=""
      type="submit"
    >
      Generuj fiszki
    </button>
  </form>
</div>
`;
